common/nmea_parser.h

```diff
@@ -0,0 +1,205 @@
+// NMEA message parser
+
+/*
+          1       2      3 4        5 6 7  8   9  10 11  12 13 14   15
+|         |       |      | |        | | |  |   |   | |   | |   |    |
+$--GGA,hhmmss.ss,llll.ll,a,yyyyy.yy,a,x,xx,x.x,x.x,M,x.x,M,x.x,xxxx*hh
+
+1) Time (UTC)
+2) Latitude
+3) N or S (North or South)
+4) Longitude
+5) E or W (East or West)
+6) GPS Quality Indicator,
+0 = Invalid, 1 = Valid SPS, 2 = Valid DGPS, 3 = Valid PPS
+7) Number of satellites in view, 00 - 12
+8) Horizontal Dilution of precision, lower is better
+9) Antenna Altitude above/below mean-sea-level (geoid)
+10) Units of antenna altitude, meters
+11) Geoidal separation, the difference between the WGS-84 earth
+ellipsoid and mean-sea-level (geoid), '-' means mean-sea-level below ellipsoid
+12) Units of geoidal separation, meters
+13) Age of differential GPS data, time in seconds since last SC104
+type 1 or 9 update, null field when DGPS is not used
+14) Differential reference station ID, 0000-1023
+15) Checksum
+*/
+
+#pragma once
+// standard includes
+#include<string>
+#include<vector>
+#include <sstream>
+
+// BoB includes
+#include "../third_party/units.h"
+#include "map_coordinate.h"
+
+namespace BoBRobotics
+{
+namespace GPS 
+{
+
+class GPSError
+  : public std::runtime_error
+{
+public:  
+    GPSError(const std::string &msg) : std::runtime_error(msg) { }
+};
+
+enum class GPSQuality
+{
+    INVALID = 0,  
+    GPSFIX  = 1,  
+    DGPSFIX = 2,
+    PPSFIX  = 3,
+    RTK     = 4,
+    FRTK    = 5
+};
+
+struct TimeStamp
+{
+    units::time::hour_t hour;
+    units::time::minute_t minute;
+    units::time::second_t second;
+    units::time::millisecond_t millisecond;
```

Comment:
Aren't these double-precision times? This doesn't seem the right storage format to me - personally I'd keep them as ``int`` - I don't see how these would be at all useful in contexts we use the units library in...

Reply:
Good point.